# Auteurs en graphe
## Idée 1 :  association par période
On associe un lien entre deux auteurs si ils ont utilisé le mot tous les deux la même semaine.
### Justification
Cela pourrait montrer une proximité de difffusion...
### Algorithme
* Choisir un mot (ultérieurement : détection automatique, choisir plusieurs mots)
* Rassembler toutes les lettres l'utilisant 
* Classer les lettres par semaine
* Ajouter dans un graphe d'adjacence tous les auteurs utilisant ce mot

### Résultat / Affichage
* Sous forme de graphe avec le nom du groupe ?

### Calculs supplémentaires
* Observer la présence des groupes

### Améliorations possibles
* Par période plutôt que la même semaine

### Continuité
* Observer si il y a une sorte de diffusion à travers le graphe au cours du temps

In [2]:
import pandas as pd
import datetime as dt
import matplotlib.pyplot as plt
data_path = "../data/"

In [3]:
# Importation des données utilisées
df = pd.read_pickle(data_path+"df_2017_avec_auteurs.pkl")
ref_noms = df[['nom_auteur','ref_auteur','groupe_auteur']].groupby('ref_auteur').first()

In [465]:
# On choisit d'abord une taille maximale du nombre de lettres pour les tests, mettre bcp pour sans limite
MAX_LEN = 1000

# On choisit le titre ou la question ou les deux
titre = True
question = True

# Choix du mot, ou une expression Regex arbitraire pour l'instant
reg_exp = 'covid'

In [466]:
# Construction de la liste des supports
support = []
if titre :
    support.append('titre')
if question :
    support.append('question')

In [467]:
# On rassemble toutes les lettres utilisant le mot
df_filtered = df[df[support].sum(axis=1).str.contains(reg_exp,regex=True)].reset_index().head(MAX_LEN)
print(len(df_filtered), "lettres contiennent l'expression ", reg_exp)

953 lettres contiennent l'expression  covid


In [468]:
# On sélectionne tous les auteurs contenus dans les lettres
auteurs_filtered = df_filtered[['nom_auteur','ref_auteur','groupe_auteur']].groupby('ref_auteur').first()
auteurs_filtered['nombre_lettres'] = df_filtered.groupby('ref_auteur').count()['question']
print(len(auteurs_filtered), "auteurs ont écrit une lettre contenant l'expression ", reg_exp)

240 auteurs ont écrit une lettre contenant l'expression  covid


In [469]:
# Regrouper les lettres par semaine
print(len(df_filtered.groupby('date_question').count()),"semaines différentes")

10 semaines différentes


In [453]:
# matrice d'adjacence
mat_adj = pd.DataFrame(columns = auteurs_filtered.index, index = auteurs_filtered.index).fillna(0)

# Ajouter un lien à la matrice d'adjacence à chaque fois que deux auteurs utilisent le mot la même semaine

# Fonction qui parcours une liste d'auteurs et complète la matrice
# On fait 'set' pour éviter qu'un auteur ayant un lien de 2 aie 2 liens avec tout le monde
def add_auteurs(x):
    for a1 in set(x['ref_auteur']):    
        for a2 in set(x['ref_auteur']):
            mat_adj.loc[a1,a2] +=1
    
# A chaque semaine, on applique la fonction add_auteurs sur la liste des auteurs de la semaine
df_filtered.groupby('date_question').apply(add_auteurs)
mat_adj.head(10)

ref_auteur  PA1008  PA1012  PA1206  PA1327  PA1592  PA1809  PA1838  PA1874  \
ref_auteur                                                                   
PA1008           1       1       0       1       0       0       1       1   
PA1012           1       3       2       2       0       1       2       3   
PA1206           0       2       3       2       1       1       2       3   
PA1327           1       2       2       3       1       1       2       3   
PA1592           0       0       1       1       1       0       1       1   
PA1809           0       1       1       1       0       2       0       1   
PA1838           1       2       2       2       1       0       3       3   
PA1874           1       3       3       3       1       1       3       5   
PA2086           1       2       2       2       1       0       3       4   
PA223837         0       0       1       1       1       1       1       1   

ref_auteur  PA2086  PA223837  ...  PA722170  PA722194  PA722202  PA722252  \
ref_auteur                    ...                                           
PA1008           1         0  ...         0         1         1         0   
PA1012           2         0  ...         0         2         2         0   
PA1206           2         1  ...         0         1         2         0   
PA1327           2         1  ...         0         2         2         0   
PA1592           1         1  ...         0         0         1         0   
PA1809           0         1  ...         0         1         1         0   
PA1838           3         1  ...         0         1         3         0   
PA1874           4         1  ...         1         2         4         1   
PA2086           4         1  ...         1         1         4         1   
PA223837         1         2  ...         0         0         2         0   

ref_auteur  PA722292  PA722300  PA722312  PA722390  PA942  PA998  
ref_auteur                                                        
PA1008             0         0         1         1      1      1  
PA1012             1         1         2         2      1      2  
PA1206             1         1         2         2      0      2  
PA1327             0         0         3         3      1      2  
PA1592             0         0         1         1      0      1  
PA1809             0         0         1         1      0      1  
PA1838             1         1         2         2      1      3  
PA1874             2         1         3         4      1      3  
PA2086             2         1         2         3      1      3  
PA223837           0         0         1         1      0      2  

[10 rows x 240 columns]

In [455]:
# Taille minimale des liens à afficher, nécessaire pour ne pas que ce qoit illisible et très long
MIN_SIZE = 5

# Pourcentage minimal des liens à afficher, nécessaire pour ne pas que ce qoit illisible et très long
MIN_PCT = 0.01

# Titre du graphe
title  = "Liens des utilisations sur la même semaine de l'expresssion {} pour les liens de plus de {} occurences.".format(reg_exp,MIN_SIZE)

In [456]:
# Création du graphe en .dot
fdot = open("graph.dot", "w")
fdot.write('graph{\n') # graphe non orienté
fdot.write('label="{}";\n'.format(title)) # Nom du graphe
fdot.write('labelloc="t";\n') # Localisation des labels

cpt = 0 # Compteur pour connaître l'avancement
cpt_l = 0 # Compteur des liens 
nb = len(mat_adj.columns) # Nombre total d'auteurs

# On parcours les auteurs
for i in range(nb):
    
    # Pour connaître l'avancement
    if i%(nb//10) == 0 :
        print("{}% ... ".format(round(100*cpt/nb)),end = '')
    
    a1 = mat_adj.columns[i] # Auteur 1
    nbl_1 = auteurs_filtered.loc[a1,'nombre_lettres'] # Nombre de lettres envoyées par a1
    
    # On parcours les auteurs
    for j in range(i+1,nb):
        a2 = mat_adj.columns[j]  # Auteur 2
        nbl_2 = auteurs_filtered.loc[a2,'nombre_lettres'] # Nombre de lettres envoyées par a2
        liens = (mat_adj.loc[a1,a2])
        pct_1 = liens/nbl_1 # Proportion des liens avec cet auteur
        pct_2 = liens/nbl_2 # Proportion des liens avec cet auteur 
            
        if mat_adj.loc[a1,a2]>=MIN_SIZE and (pct_1 > MIN_PCT or pct_2 > MIN_PCT):
            cpt_l+=1
            ea1 = auteurs_filtered.loc[a1]
            na1 = "{} ({}) {}".format(ea1['nom_auteur'],ea1['groupe_auteur'],nbl_1) # Nom de l'auteur
            ea2 = auteurs_filtered.loc[a2]
            na2 = "{} ({}) {}".format(ea2['nom_auteur'],ea2['groupe_auteur'],nbl_2) # Nom de l'auteur
            fdot.write('"{}"--"{}"[label="{} ({:0.2f} - {:0.2f})"]\n'.format(na1,na2,liens,pct_1, pct_2))
    cpt+=1

fdot.write('}') # Fin
fdot.close()
print("\nIl y a au total", cpt_l, "liens")

0% ... 10% ... 20% ... 30% ... 40% ... 50% ... 60% ... 70% ... 80% ... 90% ... 
Il y a au total 141 liens


In [457]:
# Créer le graph ps
!dot -Tps graph.dot -o graph.ps

## Résultats
### reg_exp = *racisme*, titre = **True**, question = **True**, MIN_SIZE = **0**, MIN_PCT = **0.0**, pas de MAX_LEN
* **45** lettres
* **32** auteurs
* **34** semaines
* **13** liens
* Calcul rapide
* Graphe lisible
* grosse proximité entre les partis

### reg_exp = *covid*, titre = **True**, question = **True**, MIN_SIZE = **5**, MIN_PCT = **0.1**, pas de MAX_LEN
* **953** lettres
* **240** auteurs
* **10** semaines
* **141** liens
* Calcul lent
* Graphe illisible